components/bl-dropzone-component/src/helpers.js

```diff
@@ -0,0 +1,97 @@
+import { useState } from 'react';
+
+const UploadStatus = {
+  success  : 'success',
+  uploading: 'uploading',
+  undefined: 'undefined',
+  error    : 'error',
+};
+
+export function useDropzone(component, eventHandlers) {
+  const { overwriteFiles, targetDirectory, uploadOnDrop } = component;
+  const { onDelete, onChange, fileNameLogic, onUpload, available } = eventHandlers;
+
+  const [files, setFiles] = useState([]);
+
+  const updateFiles = filesList => {
+    const updatedFiles = [];
+
+    filesList.forEach(async item => {
+      if (item.uploadStatus === UploadStatus.uploading) {
+        item.uploadStatus = UploadStatus.undefined;
+
+        try {
+          const file = item.file;
+          const hasUploadHandler = available('onUpload');
+
+          if (hasUploadHandler) {
+            await onUpload({ file });
+          } else {
+            const fileName = await fileNameLogic({ file });
+            const validFileName = ensureExtension(fileName, file);
+            const directory = targetDirectory?.replace(/\/$/g, '').replace(/(.*)/, '$1/') || '';
+            const path = directory + validFileName;
+
+            await Backendless.Files.upload(file, path, overwriteFiles);
+          }
+
+          item.uploadStatus = UploadStatus.success;
+        } catch (error) {
+          item.uploadStatus = UploadStatus.error;
+          console.error(error.message);
+        }
+      }
+
+      updatedFiles.push(item);
+
+      if (updatedFiles.length === filesList.length) {
```

Comment:
```
 onChange({ filesList });

for(const fileItem of filesList){
  await ...

  updatedFiles.push(fileItem);
}

setFiles(updatedFiles);
```

Reply:
If we use this method, the filesList will be updated every time the status changes inside the loop. And this will cause the updateFiles function to work incorrectly and display the wrong file upload status on the page.

So we don't have to wait for the result of the loop. Because in this case the updateFiles function will run only once and the filesList will not change during the cycle.

Reply:
hold on, in my version behaviour will be almost the same, the only difference is you won't compare `updatedFiles.length === filesList.length` and iterations will go in a synchronous manner, or did I miss something?

ok, let's get rid of the hack with status and see one more time